Readme.md
# Acacia

[![Node.js Version][node-image]][node-url]
[![NPM version][npm-image]][npm-url]
[![Dependency Status][dep-image]][dep-url]
[![build status][travis-image]][travis-url]
[![Coverage Status][cov-img]][cov-url]

Minimal Api framework based on [koa](https://github.com/koajs/koa/tree/v2.x) v2.

The goal is to have an api starter kit

## Features

- X-response-time via [response-time](https://github.com/koajs/response-time)
- log via [koa-morgan](https://github.com/koa-modules/morgan)
- cors via [koa-cors](https://github.com/evert0n/koa-cors)
- compress [koa-compress](https://github.com/koajs/compress)
- body parser via [koa-bodyparser](https://github.com/koajs/bodyparser/tree/3.x)
- Routes from directory via [koa-66-aggregate](https://github.com/menems/koa-66-aggregate)
- services helper via [services-stack](https://github.com/menems/services-stack)


## Installation

```bash
$ npm install acacia
```

## Example

take a look at [acacia-example](https://github.com/menems/acacia-example)

## Test

```bash
$ npm test

```

[node-image]: https://img.shields.io/node/v/acacia.svg?style=flat-square
[node-url]: https://nodejs.org
[npm-image]: https://img.shields.io/npm/v/acacia.svg?style=flat-square
[npm-url]: https://npmjs.org/package/acacia
[travis-image]: https://img.shields.io/travis/menems/acacia/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/menems/acacia
[cov-img]: https://coveralls.io/repos/menems/acacia/badge.svg?branch=master&service=github
[cov-url]: https://coveralls.io/github/menems/acacia?branch=master
[dep-image]: http://david-dm.org/menems/acacia.svg?style=flat-square
[dep-url]:http://david-dm.org/menems/acacia


History.md

v0.4.4 / 2016-04-20
===================

  * feat: add x-response-time

v0.4.3 / 2015-12-22
==================

  * chores: dependencies update
  * chores: dependencies update

v0.4.2 / 2015-12-07
===================

  * chores: koa-66 dependencies update

v0.4.1 / 2015-12-03
===================

  * chores: dependencies updated

v0.4.0 / 2015-12-03
===================

  * tests: more
  * tests: more
  * chores: lint
  * feat: default port
  * feat: add compress
  * feat: remove Validation
  * chores: add eslint

v0.2.1 / 2015-12-01
===================

  * fix: typo service => services

v0.2.0 / 2015-12-01
===================

  * fix: new ServicesStack
  * fix: cors wihtout origin should work

v0.1.0 / 2015-11-20
===================

  * chore: update dependencies
  * feat: remove path from config and change signature
  * feat: add Validation
  * chore: use master branch koa-66-aggregate
  * feat: add resources context support
  * chore: KoaApiBase -> Acacia


test/cors.js
'use strict';

process.env.NODE_ENV = 'test';

const request = require('supertest');
const should = require('should');
const Acacia = require('..');

let server;

describe('acacia cors', () => {
    before(done => {
        const config = {
            port: 4661,
            security: {
                cors: {
                    headers: 'Content-Type,Allow,Authorization',
                    credential: true,
                    domains: [{
                        origin: 'http://localhost',
                        methods: ['GET', 'PUT', 'HEAD', 'POST', 'DELETE']
                    }]
                }
            }
        };

        const app = new Acacia(config);
        app.resources({
            path: __dirname + '/routes'
        });
        server = app.listen(done);
    });

    after(done => {
        server.close();
        done();
    });

    it('should work without origin', done => {
        request(server)
            .get('/')
            .expect(200)
            .end((err, res) => {
                should.not.exist(err);
                res.headers['access-control-allow-origin'].should.eql('*');
                res.headers['access-control-allow-methods'].should.eql('GET,HEAD,PUT,POST,DELETE');
                res.text.should.eql('hello');
                done();
            });
    });

    it('should work with origin', done => {
        request(server)
            .get('/')
            .set('Origin', 'http://localhost')
            .expect(200)
            .end((err, res) => {
                should.not.exist(err);
                res.headers['access-control-allow-origin'].should.eql('http://localhost');
                res.headers['access-control-allow-methods'].should.eql('GET,HEAD,PUT,POST,DELETE');
                res.text.should.eql('hello');
                done();
            });
    });

    it('should not working with illegal origin', done => {
        request(server)
            .get('/')
            .set('Origin', 'pouet')
            .expect(404)
            .end(done);
    });
});


test/index.js
'use strict';

process.env.NODE_ENV = 'test';

const request = require('supertest');
const should = require('should');
const Acacia = require('..');

let server;

const config = {
    port: 4661,
    security: {
        cors: {
            headers: 'Content-Type,Allow,Authorization',
            credentials: true,
            maxAge: 1,
            domains: [{
                origin: 'http://localhost',
                methods: ['GET', 'PUT', 'HEAD', 'POST', 'DELETE']
            }]
        }
    },
    responseTime: true
};

const config2 = {
    port: 4662,
    security: {
        cors: {
            domains: [{
                origin: 'http://localhost',
                methods: ['GET', 'PUT', 'HEAD', 'POST', 'DELETE']
            }]
        }
    }
};

function createApp(config, services) {
    const app = new Acacia(config);
    app.resources({path: __dirname + '/routes'});

    if (!services) app.services({path: __dirname + '/services'});
    else app.services();

    return app.listen();
}

describe('acacia main', () => {
    it('should has default port without config', done => {
        const app = new Acacia();
        app.context.config.port.should.eql(1664);
        done();
    });

    it('should throw if services options not exist', done => {
        const app = new Acacia();
        (() => app.services()).should.throw('options is required');
        done();
    });

    it('should throw if services options is not an object', done => {
        const app = new Acacia();
        (() => app.services(1)).should.throw('options must be an object');
        done();
    });

    it('should add x-response-time if config is enable', done => {
        server  = createApp(config);
        request(server)
            .get('/')
            .expect(200)
            .expect('x-response-time', /.*ms/)
            .end(() => server.close(() => done()));
    });
});

describe('acacia cors', () => {
    it('should not set unwanted headers', done => {
        server = createApp(config2);
        request(server)
            .get('/')
            .expect(200)
            .end((err, res) => {
                should.not.exist(err);
                res.headers['access-control-allow-origin'].should.eql('*');
                res.headers['access-control-allow-methods'].should.eql('GET,HEAD,PUT,POST,DELETE');
                should.not.exist(res.headers['Access-Control-Max-Age']);
                should.not.exist(res.headers['Access-Control-Allow-Credentials']);
                res.text.should.eql('hello');
                server.close(() => done());
                // done();
            });
    });

    it('should work without origin', done => {
        server = createApp(config);
        request(server)
            .get('/')
            .expect(200)
            .end((err, res) => {
                should.not.exist(err);
                res.headers['access-control-allow-origin'].should.eql('*');
                res.headers['access-control-allow-methods'].should.eql('GET,HEAD,PUT,POST,DELETE');
                res.headers['access-control-max-age'].should.eql('1');
                res.headers['access-control-allow-credentials'].should.eql('true');
                res.text.should.eql('hello');
                server.close(() => done());
                // done();
            });
    });

    it('should work with origin', done => {
        server = createApp(config);
        request(server)
            .get('/')
            .set('Origin', 'http://localhost')
            .expect(200)
            .end((err, res) => {
                should.not.exist(err);
                res.headers['access-control-allow-origin'].should.eql('http://localhost');
                res.headers['access-control-allow-methods'].should.eql('GET,HEAD,PUT,POST,DELETE');
                res.text.should.eql('hello');
                server.close(() => done());
            });
    });

    it('should not working with illegal origin', done => {
        server = createApp(config);
        request(server)
            .get('/')
            .set('Origin', 'pouet')
            .expect(404)
            .end(() => server.close(() => done()));
    });
});


test/routes/index.js
const Router = require('koa-66');
const router = module.exports = new Router();

router.get('/', ctx => ctx.body = 'hello');


index.js
'use strict';

const Koa = require('koa');
const morgan = require('koa-morgan');
const convert = require('koa-convert');
const cors = require('koa-cors');
const bodyParser = require('koa-bodyparser');
const compress = require('koa-compress');
const resources = require('koa-66-aggregate');
const servicesStack = require('services-stack');
const extend = require('extend');
const responseTime = require('koa-response-time');

class Acacia extends Koa {


    constructor(config) {
        super();

        // expose config
        this.context.config = config || { port: 1664};

        if (this.context.config.responseTime)
            this.use(responseTime());

        // logger
        if (this.env !== 'test')
            this.use(morgan(this.context.config.log || 'dev'));

        // cors
        if (this.context.config.security && this.context.config.security.cors)
            this.initCors(this.context.config.security.cors);
        // zlib
        this.use(convert(compress()));

        // body parser
        this.use(bodyParser());
    }

    initCors(config) {
        const options = {
            origin: req => {
                if (!req.header.origin ||
                    !config.domains ||
                    !config.domains.length ||
                    config.domains[0].origin === '*')
                    return '*';

                const has = config.domains.filter(domain =>
                    req.header.origin === domain.origin
                );
                return (has.length) ? has[0].origin : false;
            }
        };

        if (config.headers) options.headers = config.headers;

        if (config.credentials) options.credentials = config.credentials;

        if (config.maxAge) options.maxAge = config.maxAge;

        this.use(convert(cors(options)));
    }

    /**
     * Initialise services throw directories using
     * services-stack package.
     *
     * Inject `context` object in services of type functions
     * Base context has already default properties:
     * - config: config object (constructor param)
     * - Validation: chain-validator objet(validation-helper)
     *
     * @param  {object} options
     * @return {object} Acacia instance
     */
    services(options) {
        if (!options)
            throw new TypeError('options is required');

        if (typeof options != 'object')
            throw new TypeError('options must be an object');

        options.context = options.context || {};

        const base = {
            config: this.context.config
        };

        options.context = extend(base, options.context);
        this.context.service = servicesStack(options).get;
        return this;
    }

    /**
     * Initialise routers throw directories
     *
     * Inject `plugins` object on each Router object
     *
     * @param  {object} options
     * @return {object} Acacia instance
     */
    resources(options) {
        this.use(resources(options).routes());
        return this;
    }

    listen(fn) {
        return super.listen(this.context.config.port, () => {
            if (this.env === 'development')
                console.log('listening on ' + this.context.config.port);
            if (fn && typeof fn == 'function') fn.call(this);
        });
    }
}

module.exports = Acacia;
